compiler/rustc_lint/src/types.rs

```diff
@@ -486,6 +508,62 @@ fn lint_literal<'tcx>(
     }
 }
 
+fn lint_nan<'tcx>(
+    cx: &LateContext<'tcx>,
+    e: &'tcx hir::Expr<'tcx>,
+    binop: hir::BinOp,
+    l: &'tcx hir::Expr<'tcx>,
+    r: &'tcx hir::Expr<'tcx>,
+) {
+    fn is_nan(cx: &LateContext<'_>, expr: &hir::Expr<'_>) -> bool {
+        let expr = expr.peel_blocks().peel_borrows();
+        match expr.kind {
+            ExprKind::Path(qpath) => {
+                let Some(def_id) = cx.typeck_results().qpath_res(&qpath, expr.hir_id).opt_def_id() else { return false; };
+
+                matches!(cx.tcx.get_diagnostic_name(def_id), Some(sym::f32_nan | sym::f64_nan))
+            }
+            _ => false,
+        }
+    }
+
+    let is_l_nan = is_nan(cx, l);
+    let is_r_nan = is_l_nan || is_nan(cx, r);
```

Comment:
```suggestion
    let is_r_nan = is_nan(cx, r);
```
`l` being nan does not imply `r` being nan.

Reply:
Yes, but if `l` is nan it doesn't not matter if `r` is nan or not, and since the computation is not free I added that little "trick".

I will add a noting what I just said.

Reply:
See https://github.com/rust-lang/rust/pull/111818/files#r1214819987 which would avoid having this trick.